src/DataValues/TimeValueCalculator.php

```diff
@@ -125,4 +185,115 @@ public function getSecondsForPrecision( $precision ) {
 		throw new InvalidArgumentException( "Unable to get seconds for precision $precision." );
 	}
 
+	/**
+	 * @param $timestamp
+	 * @param $precision
+	 *
+	 * @return string
+	 */
+	private function timestampAbsFloor( $timestamp, $precision ) {
+		// The year is padded with zeros to have 16 digits
+		$timestamp = substr_replace(
+			$timestamp,
+			str_repeat( '0', $this->MAX_LENGTH_TIMESTAMP - strlen( $timestamp ) ),
+			1,
+			0
+		);
+		$numCharsToModify = $this->charsAffectedByPrecision( $precision );
+		$timestamp = substr( $timestamp, 0, -$numCharsToModify ) .
+			substr( $this->TIMESTAMP_ZERO, -$numCharsToModify );
+		return $timestamp;
+	}
+
+	/**
+	 * @param $timestamp
+	 * @param $precision
+	 *
+	 * @return string
+	 */
+	private function timestampAbsCeiling( $timestamp, $precision ) {
+		// The year is padded with zeros to have 16 digits
+		$timestamp = substr_replace(
+			$timestamp,
+			str_repeat( '0', $this->MAX_LENGTH_TIMESTAMP - strlen( $timestamp ) ),
+			1,
+			0
+		);
+		$numCharsToModify = $this->charsAffectedByPrecision( $precision );
+		// WARNING: Day 31 will be applied to all months
+		$timestampCeiling = substr( $timestamp, 0, -$numCharsToModify ) .
+			substr( $this->HIGHEST_TIMESTAMP, -$numCharsToModify );
+		if ( $precision === TimeValue::PRECISION_MONTH ) {
+			$month = (int)substr(
+				$timestamp,
+				-$this->charsAffectedByPrecision( TimeValue::PRECISION_YEAR ),
+				2
+			);
+			$year = (int)substr(
+				$timestamp,
+				1,
+				-$this->charsAffectedByPrecision( TimeValue::PRECISION_YEAR ) - 1
+			);
+			$daysInMonth = \cal_days_in_month( CAL_GREGORIAN, $month, $year );
+			$timestampCeiling = substr( $timestamp, 0, -$numCharsToModify ) .
+				$daysInMonth .
+				substr( $this->HIGHEST_TIMESTAMP, -$numCharsToModify + 2 );
+		}
+		return $timestampCeiling;
+	}
+
+	/**
+	 * @param $precision
+	 *
+	 * @return int
+	 */
+	private function charsAffectedByPrecision( $precision ) {
```

Comment:
From what I understand this method tells me how many characters can be cut off from the end of the timestamp string. Lets say the given timestamp is `+2017-12-31T23:59:59Z`:
* To get a timestamp with 1 minute precision, you must cut 3 or 4 characters from the end.
* To get a timestamp with 1 year precision, you must cut 15 or 16 characters, which results in `+2017-` or `+2017`.
* To get a timestamp with 10 years precision, you must cut 17 characters, which results in `+201`. No option here.

I found the numbers confusing first, but can confirm them with this analysis. :-)

Maybe add some explanation to the documentation header of this function?

Reply:
This is the number of characters from the timestamp string that we should replace with the ones in $TIMESTAMP_ZERO or $HIGHEST_TIMESTAMP. In other words, the lowest number of irrelevant characters in the timestamp string (being irrelevant for exceeding the precision).

I'll add this to the documentation header. Congrats for the analysis, but you could have asked first! ;-)